Source/Core/Core/HW/Memmap.cpp

```diff
@@ -394,15 +395,48 @@ void UpdateLogicalMemory(const PowerPC::BatTable& dbat_table)
 
 void DoState(PointerWrap& p)
 {
-  bool wii = SConfig::GetInstance().bWii;
-  p.DoArray(m_pRAM, GetRamSize());
-  p.DoArray(m_pL1Cache, GetL1CacheSize());
+  const u32 current_ram_size = GetRamSize();
+  const u32 current_l1_cache_size = GetL1CacheSize();
+  const bool current_have_fake_vmem = !!m_pFakeVMEM;
+  const u32 current_fake_vmem_size = current_have_fake_vmem ? GetFakeVMemSize() : 0;
+  const bool current_have_exram = !!m_pEXRAM;
+  const u32 current_exram_size = current_have_exram ? GetExRamSize() : 0;
+
+  u32 state_ram_size = current_ram_size;
+  u32 state_l1_cache_size = current_l1_cache_size;
+  bool state_have_fake_vmem = current_have_fake_vmem;
+  u32 state_fake_vmem_size = current_fake_vmem_size;
+  bool state_have_exram = current_have_exram;
+  u32 state_exram_size = current_exram_size;
+
+  p.Do(state_ram_size);
+  p.Do(state_l1_cache_size);
+  p.Do(state_have_fake_vmem);
+  p.Do(state_fake_vmem_size);
+  p.Do(state_have_exram);
+  p.Do(state_exram_size);
+
+  // If we're loading a savestate and any of the above differs between the savestate and the current
+  // state, cancel the load. This is technically possible to support but would require a bunch of
+  // reinitialization of things that depend on these.
+  if (current_ram_size != state_ram_size || current_l1_cache_size != state_l1_cache_size ||
+      current_have_fake_vmem != state_have_fake_vmem ||
+      current_fake_vmem_size != state_fake_vmem_size || current_have_exram != state_have_exram ||
+      current_exram_size != state_exram_size)
+  {
+    Core::DisplayMessage("State is incompatible with memory settings. Aborting load state.", 3000);
+    p.SetMode(PointerWrap::MODE_VERIFY);
+    return;
+  }
+
+  p.DoArray(m_pRAM, current_ram_size);
+  p.DoArray(m_pL1Cache, current_l1_cache_size);
   p.DoMarker("Memory RAM");
-  if (m_pFakeVMEM)
-    p.DoArray(m_pFakeVMEM, GetFakeVMemSize());
+  if (current_have_fake_vmem)
```

Comment:
Maybe I'm reading this incorrectly.  Should this be `state_have_fake_vmem` ?  `current_have_fake_vmem` isn't current right, it's the past value?  At least I thought the `Do()` above pulls out the data from the save state so it'd be the new current value.

Reply:
Never mind, looks like I read the diff incorrectly.  Sorry Admiral.

Reply:
It doesn't really matter which of the two are used down here, if they were different we'd already have exited the function.